[21.05] We revamped backy's poetry2nix integration.

ctheune opened the pull request:
Due to a bug in poetry2nix some package changes (like setuptools-scm) became incompatible. To get those fixed we're now pulling in Python 3.10 and the surrounding ecosystem (poetry, poetry2nix) via our upstream 22.11 tracking repository.
This also pulls in a newer backy revision that can be used with different Ceph releases again.

@flyingcircusio/release-managers

## Release process

depends on https://github.com/flyingcircusio/backy/pull/28

Impact:

- internal backup tooling only

Changelog:

- update backy to 2.5 (unreleased):
  - Separate pool into fast an slow. According to our statistics 90% of jobs are faster than 10 minutes and after that they might run for hours. We now run the configured amount of workers both in a fast and slow pool so that long jobs may delay other long jobs but fast jobs should be able to squeeze through quickly.
  - Port to Python 3.10+.
  - Continue deleting snapshots even if some are protected.
  - Improve detection for 'whole object' RBD export: we failed to detect this is Ceph is built with wrong version options and shows itself as 'Development' instead of a real version.
  - Do not count running jobs as overdue.
  - Regularly log overdue jobs.
  - Delete Tags from former schedules
  - Add forget subcommand

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - Properly managing dependencies.
  - verify backup still works in dev environment

- [x] Security requirements tested? (EVIDENCE)
  - [x] verify this builds and backups still work in dev environment: ~~1 backup failure possibly related, investigate this PL-131339~~
  - [x] automated hydra tests still pass

ctheune committed:
We revamped backy's poetry2nix integration.

Due to a bug in poetry2nix some package changes (like setuptools-scm)
became incompatible. To get those fixed we're now pulling in Python
3.10 and the surrounding ecosystem (poetry, poetry2nix) via our upstream
22.11 tracking repository.

We also stumbled over the sqlite build: the patch was already applied -
I'm not sure why we're running into this in this change, but I had to
remove it to get it building again.

osnyx commented on pkgs/overlay.nix:
```diff
@@ -434,6 +436,12 @@ in {
   });
 
+  poetry = nixpkgs-22_11.poetry;
+  poetry2nix = (super.lib.debug.traceVal nixpkgs-22_11.poetry2nix);
```
Do we still need the tracing? If yes, why?

ctheune replied:
no we don't, good catch!

osnyx commented on pkgs/overlay.nix:
Why don't we need to patch sqlite anymore?

osnyx commented on pkgs/backy/nautilus.nix:
```diff
@@ -3,24 +3,39 @@
+    # TODO: this points to a branch head that only differs by the Nautilus-specific commits
+    # from what we use for Luminous, but not the latest `main` head as that currently
+    # fails to build in NixOS.
+    rev = "7914012a4e74dfeca597f9dcde237de5ee2f41b0";
```
Now that the 21.05 poetry2nix build is fixed, I can also rebase the Nautilus changes onto it and see whether this simplifies the packaging here.

ctheune replied:
yes, please.

ctheune committed:
Don't try to build all upstream overlay Python 3.10 packages.

osnyx commented on pkgs/backy/nautilus.nix:
```diff
@@ -3,24 +3,39 @@
+      pytest-flake8 = super.pytest-flake8.overrideAttrs (old: {
+        buildInputs = [ super.setuptools ];
+      });
+      consulate = super.consulate.overrideAttrs (old: {
```
For fc-qemu, we use a consulate built from a more recent commit than the last release, as upstream has not made a release for quite a while.
A manual test of backy on a dev host [indicates](https://github.com/flyingcircusio/backy/pull/28#issuecomment-1410777562) that now consulate 0.6 is too old for use in backy as well. Additionally, consulate currently has code that is deprecated in python-3.10, see https://github.com/osnyx/consulate/commit/f05c5b177e03c7e14e5f92c0026d5802c1348810 for a fix suggestion.

osnyx committed:
backy: 2.5.0 with support for multiple rbd versions

- pulls in the latest upstream revision of version 2.5.0
- that version again supports the ceph releases from Jewel to Nautilus
  at once
- includes switch to poetry packaging

osnyx committed:
backyserver: use the same package independent of ceph release

Thanks to the changes in backy-2.5.0, it is again possible to use the
same package together with Ceph Jewel, Luminous or Nautilus.

osnyx commented:
Note: fc.qemu also uses consulate and needs the updated fork at latest once we move it to python-3.10 as well, but this is better being taken care of separately and tested thoroughly.

osnyx committed:
fixup! We revamped backy's poetry2nix integration.

osnyx committed:
tests.channel: remove failing nix-env -qa subtest

The subtest turns out to be very brittle in NixOS-test-vms. In this
particular case, it fails due to the newly imported additional
nixpkgs-22_11 channel:

error: cannot import '/nix/store/2q3dd73ppbj5wka16dvnjqfak347xmfv-nixpkgs-22.11-6a70ea11060', since path '/nix/store/y1z2pm89yya6a5qnjsvp0d7369h3z05f-nixpkgs-22.11-6a70ea11060.drv' is not valid, at /nix/store/j0aav96lqhxl7fjz8126ixcmkzbyv6v3-fc-overlay/pkgs/overlay.nix:12:19

The derivation file for that channel was not pulled into the closure of
the VM but is required for evaluation by nix-env.

As @dpausp and I did not really manage to reliably pull that derivation
file in, we have decided to remove the subtest and replace it with a
periodic sensu check on a testvm.

osnyx commented:
This is ready for review. The fixups will be squashed once the corresponding PR in backy is merged and the revision has been bumped one last time here.

ctheune reviewed (commented):
LGTM
